6주차 과제(리액트 라우터) 리뷰 요청 드립니다.

ccf05017 opened the pull request:
안녕하세요, 코드숨 리액트 1기 수강생 조범진입니다.

한번에 과제를 다 구현하고 커밋을 왕창 보내는 흉악한 짓(?)을 방지하기 위해
아직 미완성이지만 리뷰 부탁드립니다 ㅎㅎ

현재 루트 경로(/)에서 각각의 페이지로 라우팅하는 기능 테스트 및 구현은 완료된 상태이며,
이후 레스토랑 상세 페이지 이동 기능 구현 예정입니다.

과제 진행 중 질문 사항은 현재 App 컴포넌트에서 라우팅을 담당하고,
HomePage 컴포넌트에서 다른 컴포넌트로 Link하는 기능을 담당하는데 이 부분 테스트 범위를 잘 모르겠습니다.

현재는 App에서 HomePage의 요소를 클릭하고 이동하는 방향으로 테스트 중입니다.
이 부분이 App이 아닌 HomePage 컴포넌트 테스트에 들어가야 될 거 같은데 이 부분으로 넣으면,
라우트 경로에 따라 렌더링이 바뀌는 부분이 App에 있어서 테스트가 어렵네용.. 이 부분에 대한 조언 부탁드립니다.

감사합니다!

Commit:
docs(README): add requirements

- 과제 수행을 위한 요구사항 정리

Commit:
test(RestaurantsPage): export component

- 레스토랑 목록 컴포넌트 관리 책임을 별도의 컴포넌트로 추출

Commit:
feat(RestaurantPage, App): extract component

- 라우팅 적용을 위해 레스토랑 컴포넌트 관련 책임을 별도의 RestaurantPage 컴포넌트로 추출

Commit:
test(AboutPage): render test

- About 페이지 렌더링 시 제목이 정상적으로 렌더링되는지 확인하는 테스트 추가

Commit:
feat(AboutPage): render

- About 페이지의 기본적인 내용 렌더링 구현

Commit:
feat(index): apply router

- App 컴포넌트에서 라우팅 적용 및 테스트하기 위해 BrowserRouter 적용

Commit:
test(HomePage): render test

- 메인 화면에서 렌더링해야 할 요소들 테스트 케이스 추가

Commit:
feat(HomePage): render

- 메인 화면에 필요한 요소들(제목, About 링크, Restaurants 링크) 렌더링 기능 구현

Commit:
test(App): root path

- 루트 경로에서 렌더링해야 할 대상이 정상적으로 렌더링되는지 확인하는 테스트 추가

Commit:
test(App): /about

- About 페이지 라우팅 렌더링 케이스 추가

Commit:
refactor(App): render about page

- 라우팅 경로에 따라 About 페이지 렌더링 기능 구현

Commit:
test(App): restaurants path

- 레스토랑 목록 라우팅 케이스 추가

Commit:
refactor(App): render restaurants

- 라우팅 경로에 따라 레스토랑 목록 렌더링 기능 구현

Commit:
test(HomePage): apply router test

- 라우팅이 적용될 컴포넌트이기 때문에 라우터 컴포넌트로 감싸도록 테스트 케이스 변경

Commit:
feat(HomePage): link

- 각 항목을 클릭하면 알맞는 화면 렌더링하도록 구현

Commit:
test(App): header

- 공통으로 나타나는 헤더 렌더링 케이스 추가

Commit:
feat(App): render header

- 홈으로 돌아갈 수 있는 헤더 엘레멘트 렌더링

Commit:
refactor(App.test): extract method

- 반복적으로 사용되는 App 렌더링 기능 메서드 추출

Commit:
test(App): click header

- 헤더 클릭 시 홈으로 이동하는 테스트 케이스 추가

Commit:
feat(App); click header

- 헤더 클릭 시 홈 화면으로 이동하는 기능 추가

Commit:
test(App.test): home click case

- 루트 경로에서 클릭 가능한 요소를 클릭했을 때 정상적으로 라우팅되는지 확인하는 테스트 추가

Commit:
test(RestaurantDetail): render test

- 레스토랑 객체를 props로 전달 받아서 렌더링하는 테스트 케이스 추가

Commit:
feat(RestaurantDetail): render

- 음식점명, 주소, 메뉴를 렌더링하는 컴포넌트 구현

Commit:
test(RestaurantDetailContainer): redux test

- 리덕스에서 restaurant 상태 불러오고 렌더링 하는 테스트 케이스 추가

Commit:
feat(RestaurantDetailContainer): get from redux

- 리덕스로부터 restaurant 정보를 가져와서 RestaurantDetail 컴포넌트로 전달하는 기능 구현

Commit:
test(RestaurantPage): get param from router

- 리액트 라우터에서 파라미터를 불러오고 불러온 파라미터로 restaurant 상세 조회하는지 확인하는 테스트 케이스 작성

Commit:
feat(actions): loadRestaurant

- 임시 함수 구현

Commit:
feat(RestaurantDetailPage): get param from router

- 라우터에서 경로를 받아오고 해당 경로로 필요한 데이터를 외부에 요청하는 액션 dispatch 하도록 구현

Commit:
docs(README): add reducer requirements

- reducer 구현 요구사항 추가

Commit:
test(setRestaurant): set reducer state

- 레스토랑 세부 정보 상태를 관리하는 리듀서 테스트 케이스 구현

Commit:
feat(reducer): setRestaurant

- 레스토랑 상세정보를 전달받아서 reducer 상태를 업데이트하는 리듀서 기능 구현

Commit:
feat(actions): loadRestaurant

- 전달받은 ID로 레스토랑 상세 정보를 API 요청한 뒤 상태 업데이트

Commit:
test(App): restaurant detail case

- 레스토랑 상세 정보 조회 시도 테스트 케이스 추가

Commit:
feat(App): restaurant detail

- 라우팅 경로에 따라 레스토랑 상세 페이지로 이동하는 기능 구현

Commit:
test(api): fetchRestaurant

- 외부 서버에 레스토랑 상세 정보를 요청해서 받아오는 테스트 케이스 추가

Commit:
feat(api): fetchRestaurant

- 외부 서버에 레스토랑 상세정보를 요청해서 받아오는 기능 구현

wholemann commented on src/App.test.jsx:
```diff
@@ -24,4 +26,18 @@ describe('App', () => {
     }));
   });
+
+  context('at path /', () => {
```
'with path /' 가 좋겠네요.

보통 context는 `when`, `with`, `without`으로 시작합니다.

- http://www.betterspecs.org/

wholemann commented on src/App.test.jsx:
```diff
@@ -40,4 +40,16 @@ describe('App', () => {
     });
   });
+
+  context('at path /about', () => {
```
'with path /about'

wholemann commented on src/App.test.jsx:
```diff
@@ -40,4 +40,16 @@ describe('App', () => {
   });
+
+  context('at path /about', () => {
+    it('render about', () => {
```
'renders the about page'

완결된 문장으로 해주세요~

wholemann commented on src/RestaurantDetail.test.jsx:
```diff
@@ -0,0 +1,18 @@
+
+import RestaurantDetail from './RestaurantDetail';
+
+import RESTAURANT from '../fixtures/restaurant';
```
fixture는 상수는 아니기 때문에 RESTAURANT는 적절하지 않은 것 같습니다.
소문자로 표현되는 게 적절해 보여요.

ccf05017 replied:
요 부분이 좀 헷갈리더라구요.. 
상수가 아니라 소문자로 써놓으면 나중에 제가 제 코드 보면서 '엥? 이건 뭐야 어디서 왔어?'라고 생각할 때가 생깁니다 ㅎㅎ
나중에 작성한 코드를 읽을 때 픽스처에서 가져오는 대상은 '그냥 그런가보다'하고 생각하고 테스트 동작만 본다는 관점으로 접근해야 할지, 아니면 이걸 그래도 픽스처에서 가져온 가짜라는걸 드러내기라도 해야 될지 고민되네요
수형님은 픽스처 다룰 때 헷갈리는 경우 어떻게 다루는지 공유해주실 수 있을까요 ??

wholemann replied:
테스트 의도를 드러내는데 비중을 생각해보시면 될 것 같습니다.

애초에 테스트에 넣는 데이터 자체가 모두 가짜입니다.

내부 데이터 내용을 알아야 정확한 의도가 표현되는 테스트라면 애초에 fixture로 분리해선 안 되겠죠.

wholemann commented on src/RestaurantDetail.test.jsx:
```diff
@@ -0,0 +1,18 @@
+    <RestaurantDetail restaurant={RESTAURANT} />
+  ));
+
+  expect(getByText('양천주가')).not.toBeNull();
+  expect(getByText('서울 강남구')).not.toBeNull();
+  expect(getByText('탕수육')).not.toBeNull();
+  expect(getByText('팔보채')).not.toBeNull();
```
```suggestion
  expect(container).toHaveTextContent('양천주가');
  ...
```

DOM element를 꼭 얻어야 할 상황이 아니라면 위와 같이 표현하는 게 좀 더 표현력이 좋다고 봅니다.

wholemann commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,21 @@
+
+import RestaurantDetailContainer from './RestaurantDetailContainer';
+
+export default function RestaurantDetailPage() {
+  const { restaurantId } = useParams();
```
```suggestion
export default function RestaurantDetailPage({ params }) {
  const { restaurantId } = params || useParams();
```

위와 같이 고친다면 어떨까요? 테스트에서 어떻게 변경하면 useParams를 가짜로 구현하지 않아도 될까요?

ccf05017 replied:
..?! 이 부분은 제가 학습이 부족했던 것 같습니다. 
저는 라우팅으로 들어오는 경로를 받으려면 무조건 useParams를 써야되는 줄 알았습니다. 좀 더 학습하고 개선해보겠습니다.

ccf05017 replied:
죄송합니다 제가 완전히 잘못 이해했었네요 ㅎㅎ;;
수형님 말씀주신 부분은 props가 없을 때(실제 동작할 때)는 useParams를 사용하고,
테스트에서는 해당 컴포넌트에 제가 원하는 값(useParams로 전달되기를 바라는 값)을 받아서 사용하도록 개선하는 방향을 제시해주신 것 같은데 의도하신 바가 이게 맞나요 ?_?
추가로 궁금한 점이 이런 식으로 구현 코드에 테스트를 위한 부분(실제 동작에는 영향이 없지만)을 넣어도 상관 없을까요? 유지보수할 때 저 외의 다른 사람이 보면 '엥 이건 왜 있지?' 싶을 수도 있을 거 같아서요. 수형님께서는 평소 이런 경우 어떻게 처리하시는지 궁금합니다!

wholemann replied:
전 별로 상관 없다고 생각해요. 인터페이스는 다양해도 좋다고 봅니다.

Java에서도 테스트를 위한 생성자도 허용하는 걸요.

wholemann commented on src/App.test.jsx:
```diff
@@ -1,34 +1,94 @@
+      expect(container).toHaveTextContent('Restaurants');
+    });
+
+    it('click About', () => {
```
'it click About'이란 문장이 일단 올바른 문장이 아닌 듯 하네요.

사실 앱 입장에서 라우팅이 관심사인 듯 한데 그 부분은 하단의 테스트들로 충분히 검증된 듯 합니다.

클릭은 어차피 Link에서 제공하는 당연한 기능(?)이라서 굳이 App에서 검증하기보단

E2E에서 해도 무방해 보여요.

ccf05017 replied:
이 부분을 테스트해야 되는지 말아야 되는지 고민이 많았는데 감사합니당! ㅎㅎ
요 부분도 차력(?)에 가까운 영역이었군요..
한가지 궁금한건 그렇게 되면 클릭은 당연한 기능이지만 해당 컴포넌트(여기서는 HomePage)가 Link를 제대로 갖다 쓰는지 검증을 해야 덜 불안할 거 같아서 이걸 검증하고 싶었습니다. 
이 부분은 테스트 코드에서 MemoryRouter로 감싸지 않으면 오류가 나는 걸로 검증하고 넘어가도 될지 아니면 명시적으로 Link 컴포넌트를 구현 코드에서 잘 불러오는 걸 확인해야 되는지 경계 잡기가 힘드네요 ㅜ 이 부분 조언주시면 감사하겠습니당!

wholemann replied:
저는 덜 불안하다면 하는게 맞다고 봅니다. 그 적정 수준은 개인마다 다를 것 같아요.

저라면 믿고 쓰고 E2E로 충분히 검증할 수 있는 부분이라면 굳이 안 하는 편이라서요.
(심지어 저건 손과 눈으로도 가능하죠ㅎㅎ)

wholemann reviewed (commented):
오우 잘해주셨네요.

다만 RestaurantDetailPage가 구현되기 전에

라우팅 정책에 대해서 먼저 정하고 실제 Page를 구현했으면 더 좋았을 것 같습니다.
브라우저에서 직접 클릭해서 제대로 라우팅이 되는지 확인하는 건 매우 간단하기 때문이죠.

지금은 해당 Page에 대한 라우팅이 없네요.

comment 드린 부분을 바탕으로 일단 계속 진행해주세요.

질문이 있으시면 DM으로 부탁드려요:)

Commit:
feat(RestaurantsContainer): add Link to detail

- 레스토랑 상세 페이지로 이동하기 위한 라우팅 기능 추가

Commit:
test(reducer): modify initialState

- 초기 상태에서 레스토랑 상세 페이지 불러올 때를 위해 restaurant 초기 상태를 빈 객체로 변경하는 테스트 케이스 작성

Commit:
refactor(reducer): modify restauant init state

- 레스토랑 상세 페이지 렌더링을 위해 레스토랑 초기 상태를 빈 객체로 변경

Commit:
refactor(App.test): modify test comments

- 좀 더 명확한 테스트 코드 표현을 위해 문구 수정

Commit:
refactor(App.test): modify fixture

- 픽스처를 상수 표현이 아닌 일반 변수 표현으로 변경

Commit:
refactor(RestaurantDetail.test): apply feedback

- 픽스처 표현을 일반 변수 표현으로 변경
- DOM이 아닌 렌더링 된 컴포넌트의 요소 존재 여부 확인으로 테스트 케이스 변경

Commit:
refactor(RestaurantDetailPage): apply feedback

- useParams 모킹 처리가 아닌 props를 통해서 해결하도록 개선
- 이렇게 되면 테스트 코드에서 구현 세부사항을 숨기고 테스트에만 집중할 수 있을 것 같음

Commit:
refactor(RestaurantDetailPage.test): apply feedback

- 픽스처를 일반 변수 표현으로 변경(기존: 상수 표현)

Commit:
refactor(RestaurantDetail): loading state

- 아직 데이터 로딩이 완료되지 않은 경우 '로딩중입니다...' 표기

Commit:
refactor(redcuser): modify init state

- restaurant 초기 상태 다시 null로 변경(로딩중 화면 표기를 위해)

Commit:
docs(README): update progres

- 진행상황 업데이트

Commit:
chore(webpack.config): add hitoryApiFallback

- 라우팅 경로 설정을 위해 컨피그 추가

ccf05017 commented:
안녕하세요, 코드숨 리액트 1기 수강생 조범진입니다.

기능 구현 및 피드백 반영 완료해서 코드 리뷰 다시 요청드립니다.

현재 라우팅 기능 구현 및 레스토랑 목록에서 클릭을 통해 이동하는 부분까지는 완료됐는데,
직접 url을 입력해서 '/restaurants/1'로 접속하면 404가 나옵니다...

에러 메시지에 나오는 `http://localhost:8080/restaurants/main.js` 로 봐서는 어떤 설정값을 빼먹어서
숫자 경로를 자바스크립트에서 실행하고 보내는 게 아니라 엉뚱한 곳으로 보내고 해석하는 게 못하는 거 아닌가 싶은데
해결책을 모르겠네요 ㅜ 좀 더 찾아보고 방안 찾으면 저녁까지 다시 수정하겠습니다.

혹시나 그때까지 수정되지 않는다면 이 부분에 대해서도 고칠 방안 힌트 주시면 감사하겠습니다 :)

Commit:
feat(NotFoundPage): render not found

- 존재하지 않는 라우팅 경로로 요청 시 렌더링되는 화면 구현

Commit:
debug(index.html): script source

- /restaurants/{restaurantId}로 직접 접속하는 경우 발생하는 404 처리를 위해 /main.js로 경로 변경

Commit:
test(NotFoundPage): render test

- 찾을 수 없는 페이지 렌더링 테스트 케이스 추가

wholemann commented:
> 안녕하세요, 코드숨 리액트 1기 수강생 조범진입니다.
> 
> 기능 구현 및 피드백 반영 완료해서 코드 리뷰 다시 요청드립니다.
> 
> 현재 라우팅 기능 구현 및 레스토랑 목록에서 클릭을 통해 이동하는 부분까지는 완료됐는데,
> 직접 url을 입력해서 '/restaurants/1'로 접속하면 404가 나옵니다...
> 
> 에러 메시지에 나오는 `http://localhost:8080/restaurants/main.js` 로 봐서는 어떤 설정값을 빼먹어서
> 숫자 경로를 자바스크립트에서 실행하고 보내는 게 아니라 엉뚱한 곳으로 보내고 해석하는 게 못하는 거 아닌가 싶은데
> 해결책을 모르겠네요 ㅜ 좀 더 찾아보고 방안 찾으면 저녁까지 다시 수정하겠습니다.
> 
> 혹시나 그때까지 수정되지 않는다면 이 부분에 대해서도 고칠 방안 힌트 주시면 감사하겠습니다 :)
> 
> P.S. 음......... index.html에서 스크립트 소스를 'main.js -> /main.js'로 수정하니까 위의 현상이 해결 됐습니다. 왜인지는 잘 모르겠습니다.... ㅎㅎ
> 추정 원인은 index.html에서 받을 경로를 파싱할 때 웹팩 설정 중에 자동으로 '/'를 붙여주는 뭔가가 있는데 그걸 제가 안했어서 그런가 싶습니다. 이게 맞는지 틀린지 잘 모르겠네요 ㅜㅜㅜㅜ 좀 더 공부해보겠습니다. 혹시나 관련 해서 참고할만한 자료 있다면 추천해주시면 감사하겠습니다!!
> [참고했던 자료 링크](https://blog.jimmydc.com/webpack-history-api-fallback/)

슬랙에서 논의된 내용을 참고하시면 되겠습니다.

webpack 설정에 관련이 있습니다.

main.js 를 html에 삽입해주는 녀석이 있죠.

실제 main.js가 생성되는 경로를 생각해보시면 /main.js 로 했을 때 해결되는 이유를 알 수 있습니다.

절대 경로와 상대 경로 개념도 명확히 해볼 필요가 있겠네요.

wholemann reviewed (commented):
각각 질문에 대하여 답변 드렸으니 참고하시면 좋겠습니다:)

질문주신 부분도 슬랙에 공유드렸으니 이번 기회에 한번 Webpack을 학습하시는 게 좋겠네요ㅎㅎ

문제가 생기면 일단 나에게 보여지는 현상에 주목해볼 필요가 있습니다.

main.js가 원래 어디에 있어야 정상 작동할까요?
그렇다면 지금은 문제가 왜 생긴 걸까요?

왜 `/main.js` 로 하면 문제가 해결되는지를 힌트 삼아 생각해보면 답에 도달하실 수 있습니다:)

ccf05017 commented:
피드백 감사드립니다 ㅎㅎ
프론트를 많이 접하지 않아서 웹팩을 막연하게 '뭔가 묶어주는 것' 정도로 생각하고 넘어갔는데, 살펴봐야 될 이유가 명확해진 것 같아요 ㅎㅎ
이번 기회에 공유주신 문서 보면서 학습 해보겠습니다. :)
